iati/validator.py

```diff
@@ -0,0 +1,730 @@
+"""A module containing validation functionality."""
+
+import sys
+from lxml import etree
+import yaml
+import iati.default
+import iati.resources
+
+
+class ValidationError(object):
+    """A base class to encapsulate information about Validation Errors."""
+
+    # pylint: disable=too-many-instance-attributes
+    def __init__(self, err_name, calling_locals=None):
+        """Create a new ValidationError.
+
+        Args:
+            err_name (str): The name of the error to use as a base.
+            calling_locals (dict): The dictionary of local variables from the calling scope. Obtained by calling `locals()`. Default is an empty dictionary.
+
+        Raises:
+            ValueError: If there is no base error with the provided name.
+
+        Todo:
+            Split message formatting into a child class and raise an error when variables are missing.
+
+        """
+        # have to set here to ensure each ValidationError has its own dictionary
+        if calling_locals is None:
+            calling_locals = dict()
+
+        try:
+            err_detail = get_error_codes()[err_name]
+        except (KeyError, TypeError):
+            raise ValueError('{err_name} is not a known type of ValidationError.'.format(**locals()))
+
+        # set general attributes for this type of error
+        self.name = err_name
+        self.actual_value = None
+
+        for key, val in err_detail.items():
+            setattr(self, key, val)
+
+        self.status = 'error' if err_name.split('-')[0] == 'err' else 'warning'
+
+        # format error messages with context-specific info
+        try:
+            self.help = self.help.format(**calling_locals)
+            self.info = self.info.format(**calling_locals)
+        except KeyError:  # as missing_var_err:
+            # raise NameError('The calling scope must contain a `{missing_var_err.args[0]}` variable for providing information for the error message.'.format(**locals()))
+            pass
+
+        # set general attributes for this type of error that require context from the calling scope
+        # TODO: Determine what the defaults should be should the appropriate values not be available  # pylint: disable=fixme
```

Comment:
Is this still relevant for right now?

Reply:
It's still a TODO, though requires more work than there is time for on this PR.

https://github.com/IATI/pyIATI/pull/209/commits/f7e4246398256558a379adc2691763b4ae690614 moves this and a couple of other inline TODOs to docstrings.